certmgr/mgr/file.go

```diff
@@ -0,0 +1,279 @@
+	// This specifies the remote upstream to talk to.
+	Remote string `json:"remote" yaml:"remote"`
+}
+
+// FinalizeSpecOptionParsing backfills the embedded SpecOptions structure with values parsed during unmarshall'ing.
+// This should be invoked before you pass SpecOptions to other consumers.
+func (p *ParsableSpecOptions) FinalizeSpecOptionParsing() {
+	if p.ParsedBefore != 0 {
+		p.Before = time.Duration(p.ParsedBefore)
+	}
+	if p.ParsedInterval != 0 {
+		p.Interval = time.Duration(p.ParsedInterval)
+	}
+	if p.ParsedIntervalSplay != 0 {
+		p.IntervalSplay = time.Duration(p.ParsedIntervalSplay)
+	}
+	if p.ParsedInitialSplay != 0 {
+		p.InitialSplay = time.Duration(p.ParsedInitialSplay)
+	}
+}
+
+// A ParsableSpec is an intermediate struct representing a certmgr spec
+// on disk; this is used for parsing and converted into a Spec
+type ParsableSpec struct {
+	ParsableSpecOptions
+
+	// The service is the service that uses this certificate. If
+	// this field is not empty, the action below will be applied
+	// to this service upon certificate renewal. It can also be
+	// used to describe what this certificate is for.
+	Service string `json:"service" yaml:"service"`
+
+	// Action is one of empty, "nop", "reload", or "restart" (see
+	// the svcmgr package for details).
+	Action string `json:"action" yaml:"action"`
+
+	// Request contains the CSR metadata needed to request a
+	// certificate.
+	Request *csr.CertificateRequest `json:"request" yaml:"request"`
+
+	// Key contains the file metadata for the private key.
+	Key *util.File `json:"private_key" yaml:"private_key"`
+
+	// Cert contains the file metadata for the certificate.
+	Cert *util.CertificateFile `json:"certificate" yaml:"certificate"`
+
+	CA *util.CertificateFile `json:"ca" yaml:"ca"`
+
+	// CA specifies the certificate authority that should be used.
+	Authority ParsableAuthority `json:"authority" yaml:"authority"`
+}
+
+// loadFromPath load and fill this spec from the given pathway
+// If this invocation returns an error, the spec instance should be discarded
+// and recreated.
+func (spec *ParsableSpec) loadFromPath(path string) error {
+	in, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	switch filepath.Ext(path) {
+	case ".json":
+		err = StrictJSONUnmarshal(in, &spec)
+	case ".yml", ".yaml":
+		err = yaml.UnmarshalStrict(in, &spec)
+	default:
+		err = fmt.Errorf("unrecognised spec file format for %s", path)
+	}
+
+	return err
+}
+
+// ReadSpecFile reads a spec from a JSON configuration util.
+func ReadSpecFile(path string, defaults *ParsableSpecOptions) (*cert.Spec, error) {
+	var spec = &ParsableSpec{
+		Request: csr.New(),
+	}
+	if defaults != nil {
+		spec.ParsableSpecOptions = *defaults
+	}
+	if spec.Before == 0 {
+		spec.Before = cert.DefaultBefore
+	}
+	if spec.Interval == 0 {
+		spec.Interval = cert.DefaultInterval
+	}
+
+	err := spec.loadFromPath(path)
+	if err != nil {
+		return nil, err
```

Comment:
wrapping the error message with some context may be helpful here (and in the rest of this function) as well (but not useful inside `loadFromPath()` itself

Reply:
ack'd, and updating.